File: shared/src/background.ts
import { z } from "zod";

// A small curated set of gradients, defined as color stops rather than image
// files — both the website (CSS) and Android (Compose Brush) can render a
// two-color gradient natively, so no image generation/storage is needed for
// presets. Keep this list and docs/protocol.md's copy in sync (Android
// mirrors it by hand — see data/protocol/Background.kt).
export const BACKGROUND_PRESETS = [
  { id: "midnight", name: "Midnight", colors: ["#0f172a", "#1e293b"] },
  { id: "sunset", name: "Sunset", colors: ["#f97316", "#ec4899"] },
  { id: "ocean", name: "Ocean", colors: ["#0ea5e9", "#0f172a"] },
  { id: "forest", name: "Forest", colors: ["#065f46", "#022c22"] },
  { id: "violet", name: "Violet", colors: ["#7c3aed", "#1e1b4b"] },
  { id: "graphite", name: "Graphite", colors: ["#27272a", "#09090b"] },
] as const satisfies { id: string; name: string; colors: [string, string] }[];

export function getBackgroundPreset(id: string) {
  return BACKGROUND_PRESETS.find((p) => p.id === id);
}

export const BackgroundConfigSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("none") }),
  z.object({ type: z.literal("preset"), presetId: z.string() }),
  z.object({ type: z.literal("custom"), url: z.string() }),
]);

export type BackgroundConfig = z.infer<typeof BackgroundConfigSchema>;

File: android/app/src/main/java/com/deckboard/app/data/protocol/Background.kt
package com.deckboard.app.data.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// Mirrors shared/src/background.ts — see docs/protocol.md. Keep BackgroundPresets
// in sync with BACKGROUND_PRESETS by hand whenever presets change.
@Serializable
sealed class BackgroundConfig {
    @Serializable
    @SerialName("none")
    data object None : BackgroundConfig()

    @Serializable
    @SerialName("preset")
    data class Preset(val presetId: String) : BackgroundConfig()

    @Serializable
    @SerialName("custom")
    data class Custom(val url: String) : BackgroundConfig()
}

data class BackgroundPreset(val id: String, val name: String, val colors: Pair<Long, Long>)

object BackgroundPresets {
    val all: List<BackgroundPreset> = listOf(
        BackgroundPreset("midnight", "Midnight", 0xFF0F172A to 0xFF1E293B),
        BackgroundPreset("sunset", "Sunset", 0xFFF97316 to 0xFFEC4899),
        BackgroundPreset("ocean", "Ocean", 0xFF0EA5E9 to 0xFF0F172A),
        BackgroundPreset("forest", "Forest", 0xFF065F46 to 0xFF022C22),
        BackgroundPreset("violet", "Violet", 0xFF7C3AED to 0xFF1E1B4B),
        BackgroundPreset("graphite", "Graphite", 0xFF27272A to 0xFF09090B),
    )

    fun find(id: String): BackgroundPreset? = all.firstOrNull { it.id == id }
}
